core-primitives/stf-state-handler/src/file_io.rs

```diff
@@ -18,134 +18,284 @@
-	Ok(file.write_all(b"")?)
+/// Remove a shard directory with all of its content.
+pub fn remove_shard_dir(shard: &ShardIdentifier) {
+	let shard_dir_path = shard_path(shard);
+	if let Err(e) = std::fs::remove_dir_all(&shard_dir_path) {
+		error!("Failed to remove shard directory {:?}: {:?}", shard_dir_path, e);
+	}
 }
 
-pub(crate) fn read(path: &str) -> Result<Vec<u8>> {
-	let mut bytes = io_read(path)?;
+pub(crate) fn state_file_path(shard: &ShardIdentifier, file_name: &str) -> PathBuf {
+	let mut shard_file_path = shard_path(shard);
+	shard_file_path.push(file_name);
+	shard_file_path
+}
 
-	if bytes.is_empty() {
-		return Ok(bytes)
-	}
+pub(crate) fn shard_path(shard: &ShardIdentifier) -> PathBuf {
+	PathBuf::from(format!("{}/{}", SHARDS_PATH, shard.encode().to_base58()))
+}
 
-	let state_hash = rsgx_sha256_slice(&bytes)?;
-	debug!(
-		"read encrypted state with hash {:?} from {}",
-		H256::from_slice(state_hash.as_ref()),
-		path
-	);
+#[allow(unused)]
+fn state_file_exists(shard: &ShardIdentifier, file_name: &str) -> bool {
+	state_file_path(shard, file_name).exists()
+}
 
-	AesSeal::unseal().map(|key| key.decrypt(&mut bytes))??;
-	trace!("buffer decrypted = {:?}", bytes);
+#[allow(unused)]
+pub(crate) fn shard_exists(shard: &ShardIdentifier) -> bool {
+	let shard_path = shard_path(shard);
+	if !shard_path.exists() {
+		return false
+	}
 
-	Ok(bytes)
+	shard_path
+		.read_dir()
+		// when the iterator over all files in the directory returns none, the directory is empty
+		.map(|mut d| d.next().is_some())
+		.unwrap_or(false)
 }
 
 #[allow(unused)]
-fn write_encrypted(bytes: &mut Vec<u8>, path: &str) -> Result<sgx_status_t> {
-	debug!("plaintext data to be written: {:?}", bytes);
-	AesSeal::unseal().map(|key| key.encrypt(bytes))?;
-	io_write(&bytes, path)?;
-	Ok(sgx_status_t::SGX_SUCCESS)
+pub(crate) fn init_shard(shard: &ShardIdentifier) -> Result<()> {
+	let path = shard_path(shard);
+	fs::create_dir_all(path).map_err(|e| Error::Other(e.into()))
 }
 
-pub(crate) fn encrypt(mut state: Vec<u8>) -> Result<Vec<u8>> {
-	AesSeal::unseal().map(|key| key.encrypt(&mut state))??;
-	Ok(state)
+fn list_items_in_directory(directory: &Path) -> Vec<String> {
+	let items = match directory.read_dir() {
+		Ok(rd) => rd,
+		Err(_) => return Vec::new(),
+	};
+
+	items
+		.flat_map(|fr| fr.map(|de| de.file_name().into_string().ok()).ok().flatten())
+		.collect()
 }
 
+#[allow(unused)]
 pub(crate) fn list_shards() -> Result<Vec<ShardIdentifier>> {
-	let files = match fs::read_dir(SHARDS_PATH) {
-		Ok(f) => f,
-		Err(_) => return Ok(Vec::new()),
-	};
+	let directory_items = list_items_in_directory(&PathBuf::from(SHARDS_PATH));
 	let mut shards = Vec::new();
-	for file_result in files {
-		let s = file_result?
-			.file_name()
-			.into_string()
-			.map_err(|_| Error::OsStringConversion)?
-			.from_base58()?;
-
-		shards.push(ShardIdentifier::decode(&mut s.as_slice())?);
+	for item in directory_items {
```

Comment:
I think `item` is still correct here, because it's not guaranteed to be a valid shard. But saying that, I realize that we should relax the condition below, to ignore everything that is not a valid shard identifier